### PR TITLE
Pin button in gridstack items

### PR DESCRIPTION
**Pull request by hbcarlos**
Solves #110 

Seams that the `locked` option is not working perfectly. When an item is pinned we can't move/resize it either.

pinned

**Questions:**
- How should we indicate the item is pinned or not? 🤔
- Should we modify the metadata format to save this property on disk?

### CODE REVIEW

**Comment by hbcarlos** [comment]
Metadata format changes with this PR, notebooks without attribute `locked` in metadata will be reset if they are open with the editor. This is also the reason why the test fails.

I'm not sure if this is the right way of attaching the toolbar: 
https://github.com/voila-dashboards/voila-gridstack/blob/83bf90e67d550c531c08dfd5e4624859b26dffab/packages/jupyterlab-gridstack/src/editor/gridstack/item.tsx#L42

The latest changes look like this:
locked

**Comment by fcollonval** [comment]
Nice work!

> I'm not sure if this is the right way of attaching the toolbar

I would suggest taking a approach similar to the file browser using a `create` function.

https://github.com/jupyterlab/jupyterlab/blob/98ccafa1e1b188e852fabb581d5d3f6aad38020e/packages/filebrowser/src/listing.ts#L216

https://github.com/jupyterlab/jupyterlab/blob/98ccafa1e1b188e852fabb581d5d3f6aad38020e/packages/filebrowser/src/listing.ts#L1761

It can be nice to set the grid item content as `Panel` rather than HTML node and add two children: the toolbar and the cell.

You can add a [`SingletonLayout`](https://jupyterlab.github.io/lumino/widgets/classes/singletonlayout.html) on GridStackItem and add to it the panel.

**Comment by hbcarlos** [comment]
Thanks for the advice!

I'm not sure how to implement the layout.

**Comment by hbcarlos** [comment]
On my last PR, I created a model for items and added signals to communicate changes to the grid.

The idea would be to do a refactor to improve the items and reduce the number of signals for items in `GridStackModel`, creating a single signal to track changes on items changing the `cellRemoved` signal for something like this:

https://github.com/voila-dashboards/voila-gridstack/blob/01bcd2f264b9c0268d815750520c261b9ec37747/packages/jupyterlab-gridstack/src/editor/gridstack/model.ts#L62

```TypeScript
enum ItemState {
  CLOSED,
  LOCKED,
  UNLOCKED
}

interface ItemChange {
   item: GridStackItemModel;
   change: ItemState;
}
```
```TypeScript
this._itemStateChanged = new Signal<this, ItemChange>(this);
```

What are your thoughts? Feedback is welcome!

**Review by fcollonval — COMMENTED** [review]
Thanks for pushing on this one @hbcarlos 

It looks almost good. I have a small concern about the item model life cycle and its connection to the toolbar rendering.

File: packages/jupyterlab-gridstack/src/editor/item/toolbar.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  ): void => {
+    this.update();
+  };
+
+  render(): JSX.Element {
+    return (
+      <>
+        {this._model.isLocked ? (
+          <div className="pin" onClick={() => this._model.unlock()}>
+            <unPinIcon.react height="16px" width="16px" />
```
Comment (fcollonval) [inline packages/jupyterlab-gridstack/src/editor/item/toolbar.tsx:37]: You should avoid hard coding styles (bad practice). Instead you should add a common class to all buttons and set the svg size using that class and CSS rules.

File: packages/jupyterlab-gridstack/src/editor/item/toolbar.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    state: ItemState
+  ): void => {
+    this.update();
+  };
+
+  render(): JSX.Element {
+    return (
+      <>
+        {this._model.isLocked ? (
+          <div className="pin" onClick={() => this._model.unlock()}>
```
Comment (fcollonval) [inline packages/jupyterlab-gridstack/src/editor/item/toolbar.tsx:36]: It is better to use meaningful html tag; in this case `button` instead of `div`.

**Review by fcollonval — CHANGES_REQUESTED** [review]
I went through the latest changes @hbcarlos 
I have minor comments that should be easy to tackle.

**Comment by hbcarlos** [comment]
Hey, thanks for the review! I'll change today.

**Review by fcollonval — APPROVED** [review]
Thanks @hbcarlos LGTM

**Pull request merged**